### DISCUSSION TITLE
rustic v0.7.0

### DISCUSSION THREAD

**Discussion by simonsan** [rustic-rs/rustic#1043] [Announcements]
# Dear `rustic` users,

This rustic version, brings two major innovations. On the one hand, we have integrated [OpenDAL](https://opendal.apache.org).
On the other hand, you can now spawn a WebDAV server with `rustic webdav`, which gives you access to the snapshots in your repository in the file browser.

Besides these big milestones, many additions have found their way to rustic and also many minor bugs have been fixed. Thanks a lot to all contributors, bug reporters or people just asking in the rustic discussions!

### Backend and Configuration

- Added many new backends using OpenDAL, including SFTP (on Unix) and S3. See the [config examples](https://github.com/rustic-rs/rustic/tree/main/config).
- Added the new `webdav` command.
- Rclone backend: Added options `rclone-command`, `use-password`, `rest-url`.
- Backend options can now be defined for hot and cold repository parts separately using `[repository.options-hot]` and `[repository.options-cold]`.

### Backup and Restore

- Backup scanning for size and start of the actual backup is now parallelized.
- Added `--no-scan` option to prevent scanning for the backup size.
- Added `--custom-ignorefile` option to use .gitignore-behavior with custom filenames.
- Added `--skip-identical-backup` option to skip saving the snapshot if it is identical to the parent.
- New option `--quiet` for backup and forget command.

### Repository Management

- `ls`: New option `--numeric-id` (can also be used as `--numeric-uid-gid`).
- `prune`: New option `--early-delete-index` allows running prune with full discs.
- The default for `--max-repack` has been changed to "10%".

### UI and Usability Improvements

- Help texts are now colored again. We may change that in a future release when reworking the CLI.
- Improved progress bars, e.g., for the `copy` command.

### Optimizations

- Reduced memory requirements, especially for large repositories.

### Build and Deployment

- Added a Dockerfile for building a rustic Docker image.
- Added the possibility to build an RPM package.
- Many improvements to the internal build process and CI pipeline.
- Building rustic with `self-update` and `webdav` commands is now an optional feature.

### Bug Fixes

- Fixed repository errors after entering a password, which now show up instead of re-prompting for the password.
- Fixed an issue where the `backup` command could abort in the unlikely case of a time shift during backup (e.g., caused by NTP syncing).
- Fixed missing cache files printing a warning.
- Fixed `repair index` wrongly setting `to-delete` for newly read pack files.
- Allowed `diff` to use `local:` in path syntax, enabling comparison with Windows paths.
- Fixed `init` not working as expected for hot/cold repositories due to a regression.
- Made output from config file handling identical to other log messages.

<hr /><em>This discussion was created from the release <a href='https://github.com/rustic-rs/rustic/releases/tag/v0.7.0'>rustic v0.7.0</a>.</em>

**Comment by kapitainsky** [comment]
It is fantastic release with OpenDAL integration - will test as soon as I can install.

Maybe it is still delayed but trying to install following official  method - https://github.com/rustic-rs/rustic?tab=readme-ov-file#installation

```
$ cargo install rustic-rs
    Updating crates.io index
     Ignored package `rustic-rs v0.6.1` is already installed, use --force to override
```

```
$ cargo install --git https://github.com/rustic-rs/rustic.git rustic-rs
    Updating git repository `https://github.com/rustic-rs/rustic.git`
  Installing rustic-rs v0.7.0 (https://github.com/rustic-rs/rustic.git#f06574be)
    Updating crates.io index
error: failed to compile `rustic-rs v0.7.0 (https://github.com/rustic-rs/rustic.git#f06574be)`, intermediate artifacts can be found at `/var/folders/36/02d6s6313zq5k3mfrhg4d8zc0000gn/T/cargo-installlSFq16`

Caused by:
  failed to select a version for `rustic_backend`.
      ... required by package `rustic-rs v0.7.0 (/Users/kptsky/.cargo/git/checkouts/rustic-7e057823cb08a6d3/f06574b)`
  versions that meet the requirements `^0.1` are: 0.1.1

  the package `rustic-rs` depends on `rustic_backend`, with features: `tokio` but `rustic_backend` does not have these features.
   It has an optional dependency with that name, but that dependency uses the "dep:" syntax in the features table, so it does not have an implicit feature with that name.


  failed to select a version for `rustic_backend` which could resolve this conflict

```

I can not rule out that I am doing something wrong - not very familiar with `rust` environment. However I would expect that suggested installation methods work. Or some pre-quietists of caveats are listed

**Reply by simonsan** [reply]
The release is being pushed to crates.io at a later time, the recommended installation method is to use the attached binaries to the release for your operating system instead of building from source. You can do that, but it's sometimes not as straight forward and will get harder, whenever we add platform dependent stuff that needs extra configuration.

**Reply by kapitainsky** [reply]
Thank you for clarification. Will download binary from release then.

**Reply by aawsome** [reply]
`cargo install rustic-rs` should now work.

**Reply by kapitainsky** [reply]
```
$ date
Sun Feb  4 07:12:43 GMT 2024

$ cargo install rustic-rs
    Updating crates.io index
  Installing rustic-rs v0.7.0
    Updating crates.io index
error: failed to compile `rustic-rs v0.7.0`, intermediate artifacts can be found at `/var/folders/36/02d6s6313zq5k3mfrhg4d8zc0000gn/T/cargo-installQfXaGa`

Caused by:
  failed to select a version for `rustic_backend`.
      ... required by package `rustic-rs v0.7.0`
  versions that meet the requirements `^0.1` are: 0.1.1

  the package `rustic-rs` depends on `rustic_backend`, with features: `tokio` but `rustic_backend` does not have these features.
   It has an optional dependency with that name, but that dependency uses the "dep:" syntax in the features table, so it does not have an implicit feature with that name.


  failed to select a version for `rustic_backend` which could resolve this conflict
```

**Reply by kapitainsky** [reply]
BTW `binstall` is broken too:

```
$ cargo binstall rustic-rs
 WARN Failed to retrieve token from `gh auth token` err=Os { code: 2, kind: NotFound, message: "No such file or directory" }
 WARN Failed to read git credential file
 INFO resolve: Resolving package: 'rustic-rs'
 WARN Failed to render url for Context {
    name: "rustic-rs",
    repo: Some(
        "https://github.com/rustic-rs/rustic",
    ),
    target: "x86_64-apple-darwin",
    version: "0.7.0",
    archive_format: Some(
        "tgz",
    ),
...
}: Failed to render template: missing key `bin`
 WARN The package rustic-rs v0.7.0 will be installed from source (with cargo)
Do you wish to continue? yes/[no]
?
```

which after choosing yes brings all to square one as `cargo install` does not work neither.

**Reply by aawsome** [reply]
Thanks for replying @kapitainsky! There is indeed no feature `tokio` in `rustic_backend`. I however have no clue why cargo would like to get this feature. In Cargo.toml I don't find anything. @simonsan do you have an idea what's going wrong here?

**Reply by simonsan** [reply]
```console
cargo install rustic-rs
    Updating crates.io index
  Downloaded rustic-rs v0.7.0
  Downloaded 1 crate (72.8 KB) in 0.71s
  Installing rustic-rs v0.7.0
    Updating crates.io index
  Downloaded time v0.3.34
  Downloaded miniz_oxide v0.7.2
  Downloaded 2 crates (174.2 KB) in 1.00s
   Compiling time v0.3.34
   Compiling miniz_oxide v0.7.2
   Compiling simple_asn1 v0.6.2
   Compiling dav-server v0.5.8
   Compiling flate2 v1.0.28
   Compiling simplelog v0.12.1
   Compiling self_update v0.39.0
   Compiling jsonwebtoken v9.2.0
   Compiling reqsign v0.14.7
   Compiling rustic_core v0.2.0
   Compiling opendal v0.44.2
   Compiling rustic_backend v0.1.1
   Compiling rustic-rs v0.7.0
    Finished release [optimized] target(s) in 5m 19s
  Installing C:\Users\dailyuse\.cargo\bin\rustic.exe
   Installed package `rustic-rs v0.7.0` (executable `rustic.exe`)
```

Can't reproduce on Win11 and OpenSuse 🤷🏽 Seems to be an issue tied to macOS (@kapitainsky if I remember right?).

**Reply by kapitainsky** [reply]
yeap - macOS here...

It always worked and only stopped now with `rustic-rs v0.7.0`

**Reply by simonsan** [reply]
> yeap - macOS here...
> 
> It always worked and only stopped now with `rustic-rs v0.7.0`

That's weird. Could you delete your `CARGO_TARGET_DIR` folder once completely and try again? Just so we can make sure this is no caching problem with some dependencies?

**Reply by kapitainsky** [reply]
Actually what I did was to remove all `rust` toolchain , caches etc. and installed it from the scratch again - now all works. Sorry for raising false alarm as clearly it was something with my local configuration. Thanks for patience.

Now I can start testing this release:)

**Reply by aawsome** [reply]
> Sorry for raising false alarm as clearly it was something with my local configuration. Thanks for patience.

No - thanks a lot for posting issues! This helps making rustic better. And if it turns out it's something outside of rustic, it may still help others getting into similar trouble...

**Comment by kapitainsky** [comment]
One more comment re releases. At the moment it contains all files together: binary, hashes and sig files:

```
rustic-v0.7.0-aarch64-apple-darwin.tar.gz
rustic-v0.7.0-aarch64-apple-darwin.tar.gz.asc
rustic-v0.7.0-aarch64-apple-darwin.tar.gz.sha256
rustic-v0.7.0-aarch64-apple-darwin.tar.gz.sig 
...

rustic-v0.7.0-x86_64-unknown-linux-musl.tar.gz
rustic-v0.7.0-x86_64-unknown-linux-musl.tar.gz.asc
rustic-v0.7.0-x86_64-unknown-linux-musl.tar.gz.sha256
rustic-v0.7.0-x86_64-unknown-linux-musl.tar.gz.sig
```

IMO it is messy and hard to find what one is looking for. Wouldn't it make sense to put all hashes in one file accompanied by one signature? It would make release downloads much more clear. This is what I see many other projects are doing (e.g. [rclone](https://github.com/rclone/rclone/releases)).

```
rustic-v0.7.0-aarch64-apple-darwin.tar.gz
...
rustic-v0.7.0-x86_64-unknown-linux-musl.tar.gz
...
rustic-v0.7.0.sha256
rustic-v0.7.0.asc
rustic-v0.7.0.sig
```

**Reply by simonsan** [reply]
We need to see, how that plays out with different tooling in the ecosystem. These files are being pulled in by package managers like `scoop` for example or also `cargo-binstall`, or the `self-update` command. We need to check, what delivers the highest compatibility.

Also regarding automation: When automating, you just pull in the binary file for your platform, so `binary`-`version`-`target`.tar.gz*, meaning it's more automation friendly right now, than hardcoding a second file to download that contains all the values - but yeah. That's not overly important, because you could strip the names, but still. I think it can sometimes be annoying to do so.

**Comment by johansmitsnl** [comment]
I want to use the new sftp backend. But the sftp service (Hetzner) uses a different port (23).

I get the error: `error: backend sftp Backend cannot be loaded: invalid port number`
The config is: `repository = "sftp:xxxxxxx@xxxxxxx.your-storagebox.de:rustic/personal"` and I also tried: `repository = "sftp:xxxxxxx@xxxxxxx.your-storagebox.de:23:rustic/personal`

**Reply by kapitainsky** [reply]
Can be teething troubles with new features.

maybe try the same way as we did for S3 Minio - https://github.com/rustic-rs/rustic/discussions/1044

use opendal backend directly `repository = "opendal:sftp"` and then specify all params in `[repository.options]` section.

It is opendal behind `repository = "sftp:` anyway.

**Reply by johansmitsnl** [reply]
That worked.

**Reply by aawsome** [reply]
I opened #1047 - indeed, currently the port is not used if a custom one is specified.
@johansmitsnl Can you confirm that the workaround I specified is working?

**Reply by johansmitsnl** [reply]
@aawsome I can confirm this works:

```toml
[repository]
password = "XXXXXX"
repository = "opendal:sftp"

[repository.options]
endpoint = "ssh://XXXXX.your-storagebox.de:23"
user = "XXXXX"
key = "/root/.ssh/id_XXXXX_ed25519"
```

**Comment by will-molloy** [comment]
Is there an example config using AWS S3 Glacier Deep Archive?

**Reply by simonsan** [reply]
Maybe you can try with

```toml
[repository]
repository = "opendal:s3"
password = "pass"

[repository.options]
root = "/"
bucket = "test-dal"
endpoint = "glacier-fips.us-east-1.amazonaws.com"
region = "us-east-1"
access_key_id = "xxx"
secret_access_key = "xxx"
```

combining with https://docs.aws.amazon.com/general/latest/gr/glacier-service.html 

and give feedback on if it works? :)

**Reply by aawsome** [reply]
see https://github.com/rustic-rs/rustic/pull/1048#discussion_r1477711094

@will-molloy if you are willing to try it out:
- use a hot/cold repository
- `endpoint` doesn't need to be set
- instead, for the cold repo part use 
```
[repository.options-cold]
default-storage-class = "DEEP_ARCHIVE" # or "GLACIER" or "GLACIER_IR"
```

Once you have a working configuration, please share it by opening a PR.

If you have questions, please don't hesitate to ask!

**Comment by will-molloy** [comment]
Just an FYI. Since this version, for my Windows backups I had to change the paths with an additional escape char.

E.g. `source = "C:\\Users\\X"`, is now `source = "C:\\\\Users\\\\X"`
